contracts/libraries/UserLib.sol

```diff
@@ -349,59 +352,86 @@ library UserLib {
+    RewardLib.PeriodRating storage periodRating = communityUser.userPeriodRewards[currentPeriod].periodRating[communityId];
+
     RewardLib.WeekReward storage weekReward = userContext.weekRewardContainer.weekReward[currentPeriod];
     if (pastPeriodsCount == 0 || communityUser.rewardPeriods[pastPeriodsCount - 1] != currentPeriod) {
       weekReward.activeUsersInPeriod.push(userAddr);
       communityUser.rewardPeriods.push(currentPeriod);
+      pastPeriodsCount++;
     }
 
-    if (!userPeriodRewards.periodRating[communityId].isActive) {
-      userPeriodRewards.periodRating[communityId].isActive = true;    // need?
+    bool isFirstTransactionInPeriod;
+    if (!periodRating.isActive) {
+      periodRating.isActive = true;
+      isFirstTransactionInPeriod = true;
       userPeriodRewards.rewardCommunities.push(communityId);
     }
 
-    int32 ratingToReward = userPeriodRewards.periodRating[communityId].ratingToReward;
-    int32 ratingToRewardChange = 0;
-
-    if (pastPeriodsCount > 0) {
-      uint16 previousWeekNumber = communityUser.rewardPeriods[pastPeriodsCount - 1]; // period now
-
-      int32 paidOutRating = userRating.payOutRating - ratingToReward;
-      
-      // If current week rating is smaller then past week reward then use it as base for the past week reward.
-      int32 previousWeekRating = communityUser.userPeriodRewards[previousWeekNumber].periodRating[communityId].rating;
-      int32 baseRewardRating =
-        CommonLib.minInt32(previousWeekRating, newRating);
-      
-      ratingToRewardChange =
-        (baseRewardRating - paidOutRating) -
-        ratingToReward;  // equal user_week_rating_after_change -
-                         // pay_out_rating;
-      
-      // If current preiod rating drops to negative reward then rating to award for current period should be 0
-      if (ratingToRewardChange + ratingToReward < 0)
-        ratingToRewardChange = -ratingToReward;
+    // pastPeriodsCount = communityUser.rewardPeriods.length; // ?? rewrite
+
+    int32 lastRatingToReward = periodRating.ratingToReward;
+    int32 lastRatingToRewardPreviousPeriod;
+    if (pastPeriodsCount == 1) {
+      periodRating.ratingToReward += rating;
+
+    } else {
+      uint16 previousPeriod = communityUser.rewardPeriods[pastPeriodsCount - 2];
+      RewardLib.PeriodRating storage previousPeriodRating = communityUser.userPeriodRewards[previousPeriod].periodRating[communityId];
+      lastRatingToRewardPreviousPeriod = previousPeriodRating.ratingToReward;
+
+      if (previousPeriod == currentPeriod - 1) {
+        if (previousPeriodRating.ratingToReward < 0) {
+          previousPeriodRating.ratingToReward += rating;
+          if (previousPeriodRating.ratingToReward > 0) {
+            periodRating.ratingToReward = previousPeriodRating.ratingToReward;
+            previousPeriodRating.ratingToReward = 0;
+          }
+        } else {
+          if (rating > 0) {
+            periodRating.ratingToReward += rating;
```

Comment:
Should it add rating or only rating change? What happens if receive upvote twice?